File: src/exercise2/go/backend.go
package main

import (
	"fmt"
	"sync"
	"time"
)

type Backend struct {
	db Database
}


func (b Backend) onAddClick(userId int, month time.Month) {
	valueOfSingleClick := 0.001
	b.db.addUserTransaction(userId, month, valueOfSingleClick)

	//recalculate balance
	balance := b.db.getUserBalance(userId)
	balance += valueOfSingleClick
	b.db.setUserBalance(userId, balance)
}

func (b Backend) onAddClickConcurrent(userId int, month time.Month, wg *sync.WaitGroup) {
	valueOfSingleClick := 0.001
	b.db.addUserTransaction(userId, month, valueOfSingleClick)

	//recalculate balance
	balance := b.db.getUserBalance(userId)
	balance += valueOfSingleClick
	b.db.setUserBalance(userId, balance)
	wg.Done()
}

func sum(transactions []float64) (total float64) {
	for _, transaction := range transactions {
		total += transaction
	}
	return
}

func (b Backend) cashOut(userId int, month time.Month) {
	balance := b.db.getUserBalance(userId)

	if balance > 0 {
		transactions := b.db.getUserTransactions(userId, month)
		amountToBePaid := sum(transactions)

		b.makePayment(month, userId, amountToBePaid)

		//recalculate balance
		balance := b.db.getUserBalance(userId)
		balance -= amountToBePaid
		b.db.setUserBalance(userId, balance)
	}
}

func (b Backend) makePayment(month time.Month, userId int, amount float64)  {
	//Some logic to do actual payment
	balance := b.db.getUserBalance(userId)
	fmt.Printf(
		"Payment made (%.3s): userId=%d, amount=%.3f, balance_before_payment=%.3f, balance_after_payment=%.3f\n",
		month, userId, amount, balance, balance - amount);
}

File: src/exercise2/java/Backend.java
import java.time.Month;
import java.util.List;

class Backend {
    private Database db = new Database();

    void onAddClick(int userId, Month month) {
        double valueOfSingleClick = 0.001;
        db.addUserTransaction(userId, month, valueOfSingleClick);

        //recalculate balance
        double balance = db.getUserBalance(userId);
        balance += valueOfSingleClick;
        db.setUserBalance(userId, balance);
    }

    void cashOut(int userId, Month month) {
        double balance = db.getUserBalance(userId);
        if (balance > 0) {
            List<Double> transactions = db.getUserTransactions(userId, month);
            Double amountToBePaid = transactions.stream().reduce(0.0, Double::sum);

            this.makePayment(month, userId, amountToBePaid);

            //recalculate balance
            balance = db.getUserBalance(userId);
            balance -= amountToBePaid;
            db.setUserBalance(userId, balance);
        }
    }

    private void makePayment(Month month, int userId, Double amount) {
        Double balance = db.getUserBalance(userId);
        System.out.printf(
                "Payment made (%.3s): userId=%d, amount=%.3f, balance_before_payment=%.3f, balance_after_payment=%.3f\n",
                month, userId, amount, balance, balance - amount);
    }
}
